File: src/ds-visuals/queue/code/queue.c
#include<stdio.h>
#include<stdlib.h>
#include<stdarg.h>
#define MAX 3
//queues using arrays.

typedef struct q{
	int arr[MAX];
	int front;
	int rear;
} Queue;

void enqueue(Queue* queue, int element){
	if(queue->rear == MAX-1)
		printf("Error- Queue Full\n");
	else
		queue->arr[++(queue->rear)] = element;
	if(queue->rear == 0)
		queue->front = 0;
	return;	
}

int dequeue(Queue*queue){
	if(queue->front == -1)
		printf("Error- Queue empty\n");
	else{
		int out = queue->arr[(queue->front)++];
		if(queue->front > queue->rear){
			queue->front = -1;
			queue->rear = -1;
		}
		return out;	
	}
	return -1;
}

void display(Queue *queue){
	int i;
	printf("queue: ");
	for(i = queue->front; i <= queue->rear; i++)
		printf("%d ",queue->arr[i]);
	printf("\n");
	return;
}

Queue createQueue(int num_elements, ...){
	Queue queue;
	int i;
	va_list list;
	queue.front = -1;
	queue.rear = -1;
    va_start(list, num_elements); 
    for (i = 1; i <= num_elements; i++) 
        enqueue(&queue, va_arg(list, int)); 
    va_end(list); 
    return queue;  
}

int main(){
	Queue queue = createQueue(3, 100, 200, 300);
	display(&queue);
	printf("element dequeued: %d\n",dequeue(&queue));
	display(&queue);
	enqueue(&queue, 400);
	display(&queue);
	return 0;
}

File: src/ds-visuals/tree/code/queue.py
class Queue(object):
	def __init__(self):
		super(Queue, self).__init__()
		self.rear = -1
		self.front = -1
		self.MAX = 10
		self.queue = [None]*self.MAX

	def enqueue(self,element):
		if(self.rear == self.MAX - 1):
			print("Error- queue full")
		else:
			self.rear += 1
			self.queue[self.rear] = element
		if self.rear == 0:
			self.front = 0;
		return;

	def dequeue(self):
		if(self.rear == -1):
			print("Error- queue empty")
		else:
			out = self.queue[self.front]
			self.front += 1
			if(self.front > self.rear):
				self.front = self.rear = -1;
			return out
		return -1

	def display(self):
		print('queue: ', end="")
		for i in range(self.front, self.rear+1):
			print(str(self.queue[i])+" ", end="")
		print()

	def createQueue(self, *elements):
		for element in elements:
			self.enqueue(element)

	def isEmpty(self):
		if(self.front == -1 or self.front > self.rear):
			return True
		return False
